### PR TITLE
Bignums as a separate opam package

### PR DESCRIPTION
**Pull request by letouzey**
Not fully ready yet (but no real showstopper either), this PR is to mark my intention to migrate Bignums (Int31, BigN, BigZ, BigQ) out of Coq repository, and place them instead in a dedicated opam package, ideally as soon as Coq 8.7.

## Why ? ##

  - Relatively small audience for this library (hardcore computationists ;-)
  - Important compilation overhead on the stdlib
  - First step toward a more component-based and customizable standard library

## Already done ##

  - Regrouped all relevant files to a same location (currently `plugins/bignums`)
  - Created a dedicated Makefile (based on coq-makefile)
 
## Still to be done ##

  - Create and populate a separate git repository with the content of `plugins/bignums`
  - Corresponding opam package
  - Identify and adapt third-party developments that use bignums
  - Could maybe the "retroknowledge" code about Int31 come someday as a plugin ?

### CODE REVIEW

**Comment by ejgallego** [comment]
Once you have the separate repository maybe you would like to update the `.travis.yml` file too.

**Comment by letouzey** [comment]
Indeed

**Comment by maximedenes** [comment]
> Could maybe the "retroknowledge" code about Int31 come someday as a plugin ?

As I mentioned two working groups ago, I'm working on rebasing primitive integers on trunk. I'm afraid this work goes a bit against making the retroknowledge modular, since it adds an `Int` constructor to terms. It also switches from 31 to 63 bit integers.

However, it is still doable to ship separately the libraries built on top of it. The treatment of 63-bit integers themselves is fairly lightweight.

**Comment by silene** [comment]
> Could maybe the "retroknowledge" code about Int31 come someday as a plugin?

Precisely because there is no way in the foreseeable future to move the retroknowledge code out of Coq, I would suggest not to remove `Int31.v` from Coq either. All the other files are fine though.

**Comment by letouzey** [comment]
Ok for leaving in the stdlib the stuff about `Int31` (and hence the definition of cyclic types).

**Comment by maximedenes** [comment]
What is the transition strategy for user developments?

**Comment by maximedenes** [comment]
During the WG discussion, we agreed to merge this. We were worried about the backward compatibility of developments importing the OPAM package and even suggested to publish an empty package for 8.6. But @pletouzey made the remark offline that if the package compiles with both 8.6 and 8.7, no action is required to ensure backward compatibility, apart from setting the right dependency constraints in the OPAM package.

**Comment by letouzey** [comment]
An external repository for bignums code is ready here https://github.com/coq/bignums .
I'll see now how to provide an opam package out of this.
The master branch of this repository compiles with coq trunk, and there is also a v8.6 branch.
The only difference between the two is `CErrors.user_err_loc` vs. `CErrors.user_err` in the syntax plugin. My dear compatibility zealots @maximedenes @ejgallego, how are we expected to provide
code that works both with 8.6 and 8.7 if there isn't any compat layer or deprecation delay for changes such as these CErrors functions ?

**Comment by ejgallego** [comment]
@letouzey you are right, we should add the removed function back, sorry for that.

This change was done when we had not settled on Ocaml `4.02.3` so deprecation was not available.

Anyways, for most plugins, the only solution is indeed to have their own compat layer or to ship two different files.

**Comment by maximedenes** [comment]
> My dear compatibility zealots @maximedenes @ejgallego

:)

We mean this for scripts. Given the current shape of the plugin API, it would be counter productive to guarantee compatibility. On this particular example, I'm a bit surprised, though, I thought @ejgallego had made sure there was a way to maintain compatibility, following a comment by Guillaume.

**Comment by ejgallego** [comment]
I re-added some missing functions, but not this one in particular.

**Comment by ejgallego** [comment]
Indeed as @maximedenes points out, given the current OCaml API model and the things discussed for the next Coq versions, ML-level compatibility will be far from stellar I'm afraid.

Adding your plugin to Travis can help alleviate some of the pain. However IMO given what the release schedule is going to be we should really recommend a synchronized release model to plugin authors. That is to say: if you are a plugin author, when Coq 2019.04 comes out, you release the 2019.04 version of your plugin, then you move development to Coq 2019.10.

IMHO this is feasible now as a 6 months rate should be OK for Coq. [Note that other projects such as GCC or Linux deem 6 months too long and they release way often].

**Comment by ejgallego** [comment]
Sorry @letouzey , `user_err_loc` is there in trunk, so I am not sure I understand what the problem is?

**Comment by letouzey** [comment]
@ejgallego, indeed `user_err_loc` is there today, but the trunk repo I was using at that moment was a few days old. Anyway, since these recent changes also brought your CAst stuff, all hope of `number_syntax.ml` compatible to both 8.6 and 8.7 is now lost. Well, so be it...

**Comment by letouzey** [comment]
Rebased, and directly propose the removal of BigN, BigZ, BigQ now that we have an separate repo for them ( https://github.com/coq/bignums ).

**Comment by maximedenes** [comment]
There multiple failures. Some of them are expected (CoLoR relying on BigN), some of them easy to fix (doc), some unexpected (CompCert).

**Comment by maximedenes** [comment]
@letouzey Are you working on a fix, or should we postpone this to 8.8?

**Comment by letouzey** [comment]
@maximedenes : Working on it today...

**Comment by letouzey** [comment]
Rebased + fixed `make doc`

**Comment by letouzey** [comment]
The CompCert failure is a really minor consequence of the removal of `BigNumPrelude`, which was (ab)using Hints a lot. Since this failure is unique and easy to fix in CompCert, I'll propose a patch for it (and an overlay here in the meantime).

For devs really relying on bignums (Color, MathClasses and its dependency Formal_topology), I'll provide overlays declaring my new repository as dependency for these devs

File: dev/ci/ci-formal-topology.sh
Comment (ejgallego) [inline dev/ci/ci-formal-topology.sh]: Please use the proper variables here, and do as `math_comp` and factorize the install into its own subroutine.

**Review by ejgallego — CHANGES_REQUESTED** [review]

File: dev/ci/ci-formal-topology.sh
Comment (letouzey) [inline dev/ci/ci-formal-topology.sh]: Minute papillon, as we say in French. I'm still working precisely on that...

**Comment by letouzey** [comment]
@ejgallego : does the new version of travis stuff suits you  ?

File: dev/ci/ci-basic-overlay.sh
Comment (ejgallego) [inline dev/ci/ci-basic-overlay.sh]: minor nit, it should be `bignums_CI_BRANCH` etc... as I hope soon much of this boilerplate is handled automagically, but the name should correspond to the entry in `ci-bignums`

Comment (letouzey) [inline dev/ci/ci-basic-overlay.sh]: Done

**Comment by maximedenes** [comment]
@ejgallego I believe @letouzey is waiting for your feedback on this one.

**Review by ejgallego — APPROVED** [review]

**Comment by ejgallego** [comment]
Oh sorry I didn't see it, sure it looks fine, they were really nits.

File: dev/ci/ci-color.sh
Comment (ejgallego) [inline dev/ci/ci-color.sh]: I see your point here but I don't think we want to use source to install packages, please, define an install function similarly to what we do for math-comp.

File: dev/ci/ci-bignums.sh
Comment (ejgallego) [inline dev/ci/ci-bignums.sh]: See my comment below.

File: dev/ci/ci-color.sh
Comment (ejgallego) [inline dev/ci/ci-color.sh]: This should be eventually integrated in CoLoR, I suggest we ask Frédéric to do an 8.7 branch.

File: dev/ci/ci-formal-topology.sh
Comment (ejgallego) [inline dev/ci/ci-formal-topology.sh]: replace by `install_bignum`

File: dev/ci/ci-math-classes.sh
Comment (ejgallego) [inline dev/ci/ci-math-classes.sh]: ditto.

File: dev/ci/ci-color.sh
Comment (Zimmi48) [inline dev/ci/ci-color.sh]: Didn't we say at the WG that this would even be a backward compatible fix?

Comment (letouzey) [inline dev/ci/ci-color.sh]: Please be precise when you speak of compatibility, since it might mean many many things. Yes the above change is "backward compatible" in the way I've been asked to ensure : after applying this change, CoLoR may still compiles with Coq 8.6 ... at the condition that the new bignums package (available both for 8.6 and 8.7) is correctly installed first.

Comment (letouzey) [inline dev/ci/ci-color.sh]: @ejgallego : agreed concerning the need of a 8.7 branch of CoLoR, but the `sed` stuff should do meanwhile (no nice overlay tricks for svn, right ?). And I don't think it's worth bothering Frédéric before this PR gets merged.

File: dev/ci/ci-color.sh
Comment (letouzey) [inline dev/ci/ci-color.sh]: > I don't think we want to use source to install packages

Why not ? I personally find it cleaner this way, with stuff about bignums located in `ci-bignums.sh` rather that who knows elsewhere. But I won't fight over this. I'll implement what you ask if I've 5min today, otherwise feel free to do what you what if that gets merged.

File: dev/ci/ci-color.sh
Comment (ejgallego) [inline dev/ci/ci-color.sh]: Sure as I said below I am just adding notes so we don't forget, none of my remarks is blocking.

**Comment by letouzey** [comment]
Rebased + removed the CompCert overlay (already merged :-).

**Comment by letouzey** [comment]
@maximedenes : all green here, at last :-)

**Pull request merged**
